src/main/java/com/salesforce/dataloader/action/AbstractAction.java

```diff
@@ -128,24 +130,40 @@ public final void execute() {
 
             while (!getMonitor().isCanceled() && visit()) {}
 
-            flush();
-            // need to close up here for sync message.
-            closeAll();
-
-            final Object[] args = { String.valueOf(getVisitor().getNumberSuccesses()),
-                    getConfig().getString(Config.OPERATION), String.valueOf(getVisitor().getNumberErrors()) };
-
-            // set the monitor to done
-            if (getMonitor().isCanceled()) {
-                getMonitor().doneSuccess(getMessage("cancel", args)); //$NON-NLS-1$
-            } else {
-                getMonitor().doneSuccess(getMessage("success", args)); //$NON-NLS-1$
-            }
 
         } catch (final Exception e) {
-            handleException(e);
+            exceptions.add(e);
         } finally {
-            closeAll();
+            try{
+                flush(); //make sure we don't early abort here
```

Comment:
closeAll doesn't flush the buffers. If the visitors threw there could be unwritten lines in the buffer.

Other option is to change closeAll to flush.